File: lib/worksheet-difficulty.ts
/** Worksheet-level difficulty → per-question D ramp bands (mirrors Python). */

/** Preset ids shown in UI (Level 0–4). Legacy ids still accepted as input. */
export type WorksheetDifficultyPreset =
  | "level-0"
  | "level-1"
  | "level-2"
  | "level-3"
  | "level-4";

export type WorksheetDifficultyInput = WorksheetDifficultyPreset | number | string;

export type WorksheetDifficultySchedule =
  | "intro"
  | "gentle"
  | "moderate"
  | "steep"
  | "intense";

export type WorksheetDifficultyRamp = {
  level: WorksheetDifficultyPreset;
  label: string;
  d_min: number;
  d_max: number;
  schedule: WorksheetDifficultySchedule;
  description: string;
  slider?: number;
};

const NAMED: Record<WorksheetDifficultyPreset, WorksheetDifficultyRamp> = {
  "level-0": {
    level: "level-0",
    label: "Level 0",
    d_min: 0,
    d_max: 3,
    schedule: "intro",
    description: "Introductory band: most items at very low D.",
  },
  "level-1": {
    level: "level-1",
    label: "Level 1",
    d_min: 0,
    d_max: 8,
    schedule: "gentle",
    description: "Most items at low D with a gentle ramp.",
  },
  "level-2": {
    level: "level-2",
    label: "Level 2",
    d_min: 3,
    d_max: 20,
    schedule: "moderate",
    description: "Mid-band difficulties with a moderate ramp.",
  },
  "level-3": {
    level: "level-3",
    label: "Level 3",
    d_min: 8,
    d_max: 25,
    schedule: "steep",
    description: "High-band difficulties with a steep ramp.",
  },
  "level-4": {
    level: "level-4",
    label: "Level 4",
    d_min: 20,
    d_max: 25,
    schedule: "intense",
    description: "Sustained high-D practice near the top of the band.",
  },
};

/**
 * Legacy / prior-preset aliases → Level ids (API / old deep links only).
 *
 * | Alias | Maps to | Notes |
 * |-------|---------|-------|
 * | easy, e | level-1 | former easy → D 0–8 |
 * | medium, m | level-2 | former medium ≈ mid band |
 * | hard, h | level-3 | former hard ≈ high band |
 * | d0-8 | level-1 | exact former preset |
 * | d4-14 | level-2 | closest mid band (was 4–14) |
 * | d10-22 | level-3 | closest high band (was 10–22) |
 * | d0-3, d3-20, d8-25, d20-25 | level-0…4 | range-id synonyms |
 */
const LEGACY_ALIASES: Record<string, WorksheetDifficultyPreset> = {
  easy: "level-1",
  e: "level-1",
  medium: "level-2",
  m: "level-2",
  hard: "level-3",
  h: "level-3",
  "d0-8": "level-1",
  "d4-14": "level-2",
  "d10-22": "level-3",
  "d0-3": "level-0",
  "d3-20": "level-2",
  "d8-25": "level-3",
  "d20-25": "level-4",
};

const SCHEDULE_BY_BUCKET: WorksheetDifficultySchedule[] = [
  "intro",
  "gentle",
  "moderate",
  "steep",
  "intense",
];

function lerp(a: number, b: number, t: number): number {
  return a + (b - a) * t;
}

/** Map Level preset, legacy alias, or a 0–24 slider to { d_min, d_max, schedule }. */
export function worksheetDifficultyToRamp(
  level: WorksheetDifficultyInput,
): WorksheetDifficultyRamp {
  if (typeof level === "string") {
    const key = level.trim().toLowerCase();
    if (key in NAMED) return { ...NAMED[key as WorksheetDifficultyPreset] };
    if (key in LEGACY_ALIASES) return { ...NAMED[LEGACY_ALIASES[key]] };
    const asNum = Number(key);
    if (!Number.isFinite(asNum)) {
      throw new Error(
        `Unknown worksheet difficulty "${level}"; use level-0…level-4 or 0–24.`,
      );
    }
    return worksheetDifficultyToRamp(asNum);
  }

  const clamped = Math.max(0, Math.min(24, Number(level)));
  const t = clamped / 24;
  const low = NAMED["level-0"];
  const high = NAMED["level-4"];
  const d_min = Math.round(lerp(low.d_min, high.d_min, t) * 10000) / 10000;
  const d_max = Math.round(lerp(low.d_max, high.d_max, t) * 10000) / 10000;
  const bucket = Math.min(4, Math.floor(t * 5));
  const named = WORKSHEET_DIFFICULTY_PRESETS[bucket];
  const schedule = SCHEDULE_BY_BUCKET[bucket];
  return {
    level: named,
    slider: clamped,
    label: `D ${clamped}`,
    d_min,
    d_max,
    schedule,
    description: `Continuous worksheet D=${clamped} → questions ramp from ${d_min} to ${d_max} (${schedule}).`,
  };
}

export const WORKSHEET_DIFFICULTY_PRESETS: WorksheetDifficultyPreset[] = [
  "level-0",
  "level-1",
  "level-2",
  "level-3",
  "level-4",
];

/** @deprecated Use WORKSHEET_DIFFICULTY_PRESETS */
export const WORKSHEET_DIFFICULTY_LEVELS = WORKSHEET_DIFFICULTY_PRESETS;

/** @deprecated Use WorksheetDifficultyPreset */
export type WorksheetDifficultyLevel = WorksheetDifficultyPreset;

export function worksheetDifficultyLabel(level: WorksheetDifficultyInput): string {
  return worksheetDifficultyToRamp(level).label;
}

/** UI option text: e.g. "Level 2 (D 3–20)". */
export function worksheetDifficultyOptionLabel(
  level: WorksheetDifficultyInput,
): string {
  const ramp = worksheetDifficultyToRamp(level);
  if (ramp.slider !== undefined) return ramp.label;
  return `${ramp.label} (D ${ramp.d_min}–${ramp.d_max})`;
}

File: question_engine/worksheet_difficulty.py
"""Worksheet-level difficulty → per-question D ramp bands.

Named Level presets (and a continuous 0–24 slider) map to ``d_min`` / ``d_max``
used by progressive practice (``difficulty_ramp`` / ``build_progressive_sections``).

Documented mapping
------------------
| Worksheet preset | label   | d_min | d_max | schedule |
|------------------|---------|-------|-------|----------|
| level-0          | Level 0 | 0     | 3     | intro    |
| level-1          | Level 1 | 0     | 8     | gentle   |
| level-2          | Level 2 | 3     | 20    | moderate |
| level-3          | Level 3 | 8     | 25    | steep    |
| level-4          | Level 4 | 20    | 25    | intense  |

There is no Level 5 — only five ranges (0–4) were specified.

Legacy / prior-preset aliases (input only — never used as display labels)
-----------------------------------------------------------------------
| Alias                         | Maps to  |
|-------------------------------|----------|
| easy, e                       | level-1  |
| medium, m                     | level-2  |
| hard, h                       | level-3  |
| d0-8                          | level-1  |
| d4-14                         | level-2  |
| d10-22                        | level-3  |
| d0-3, d3-20, d8-25, d20-25    | level-0…4 |

A numeric worksheet D in ``[0, 24]`` linearly interpolates between the level-0
and level-4 bands (so D≈0 ≈ level-0, D≈24 ≈ level-4).
"""

from __future__ import annotations

from typing import Any

WorksheetDifficulty = str | int | float

NAMED_RAMPS: dict[str, dict[str, Any]] = {
    "level-0": {
        "level": "level-0",
        "label": "Level 0",
        "d_min": 0.0,
        "d_max": 3.0,
        "schedule": "intro",
        "description": "Introductory band: most items at very low D.",
    },
    "level-1": {
        "level": "level-1",
        "label": "Level 1",
        "d_min": 0.0,
        "d_max": 8.0,
        "schedule": "gentle",
        "description": "Most items at low D with a gentle ramp.",
    },
    "level-2": {
        "level": "level-2",
        "label": "Level 2",
        "d_min": 3.0,
        "d_max": 20.0,
        "schedule": "moderate",
        "description": "Mid-band difficulties with a moderate ramp.",
    },
    "level-3": {
        "level": "level-3",
        "label": "Level 3",
        "d_min": 8.0,
        "d_max": 25.0,
        "schedule": "steep",
        "description": "High-band difficulties with a steep ramp.",
    },
    "level-4": {
        "level": "level-4",
        "label": "Level 4",
        "d_min": 20.0,
        "d_max": 25.0,
        "schedule": "intense",
        "description": "Sustained high-D practice near the top of the band.",
    },
}

WORKSHEET_DIFFICULTY_PRESETS = [
    "level-0",
    "level-1",
    "level-2",
    "level-3",
    "level-4",
]

# Legacy EMH + prior D-range aliases (input only — never used as display labels).
_LEGACY = {
    "easy": "level-1",
    "e": "level-1",
    "medium": "level-2",
    "m": "level-2",
    "hard": "level-3",
    "h": "level-3",
    "d0-8": "level-1",
    "d4-14": "level-2",
    "d10-22": "level-3",
    "d0-3": "level-0",
    "d3-20": "level-2",
    "d8-25": "level-3",
    "d20-25": "level-4",
}
for _alias, _target in _LEGACY.items():
    NAMED_RAMPS[_alias] = NAMED_RAMPS[_target]

_SCHEDULE_BY_BUCKET = ("intro", "gentle", "moderate", "steep", "intense")


def _lerp(a: float, b: float, t: float) -> float:
    return a + (b - a) * t


def worksheet_difficulty_to_ramp(level: WorksheetDifficulty) -> dict[str, Any]:
    """Map a worksheet difficulty label or 0–24 slider value to a D ramp.

    Returns ``{level, label, d_min, d_max, schedule, description}``.
    """
    if isinstance(level, str):
        key = level.strip().lower()
        if key in NAMED_RAMPS:
            return dict(NAMED_RAMPS[key])
        # Allow numeric strings ("12", "12.5").
        try:
            level = float(key)
        except ValueError as exc:
            raise ValueError(
                f"Unknown worksheet difficulty {level!r}; "
                "use level-0…level-4 or a number in [0, 24]."
            ) from exc

    t = max(0.0, min(24.0, float(level))) / 24.0
    low = NAMED_RAMPS["level-0"]
    high = NAMED_RAMPS["level-4"]
    d_min = round(_lerp(float(low["d_min"]), float(high["d_min"]), t), 4)
    d_max = round(_lerp(float(low["d_max"]), float(high["d_max"]), t), 4)
    bucket = min(4, int(t * 5))
    named = WORKSHEET_DIFFICULTY_PRESETS[bucket]
    schedule = _SCHEDULE_BY_BUCKET[bucket]
    return {
        "level": named,
        "slider": float(level) if not isinstance(level, str) else float(t * 24),
        "label": f"D {float(level):g}",
        "d_min": d_min,
        "d_max": d_max,
        "schedule": schedule,
        "description": (
            f"Continuous worksheet D={float(level):g} → "
            f"questions ramp from {d_min:g} to {d_max:g} ({schedule})."
        ),
    }


def apply_worksheet_difficulty(
    progressive: dict[str, Any],
    level: WorksheetDifficulty | None = None,
) -> dict[str, Any]:
    """Return a copy of ``progressive`` with d_min/d_max filled from worksheet difficulty.

    Uses ``level`` if given, else ``progressive["worksheet_difficulty"]``.
    Existing explicit d_min/d_max win unless ``overwrite`` is set via level only.
    """
    out = dict(progressive)
    raw = level if level is not None else out.get("worksheet_difficulty")
    if raw is None:
        return out
    ramp = worksheet_difficulty_to_ramp(raw)  # type: ignore[arg-type]
    out["worksheet_difficulty"] = ramp["level"]
    out["d_min"] = ramp["d_min"]
    out["d_max"] = ramp["d_max"]
    out["schedule"] = ramp["schedule"]
    return out
